Storage/REST-API.md

```diff
@@ -0,0 +1,183 @@
+
+### Methods
+
+Current Storage API interprets HTTP verbs in the following way:
+
+* **GET**: Get the resource.
+
+  Strongly typed with respect to returning an individual record or a collection of records. For example, `/users` can be the way to GET the list of users (`"data":[...]`), and `/user/1` can be the way to GET an individual user with the ID of `1`.
+   
+* **PATCH**: Update some fields of the resource.
+
+  `PATCH` should be called on the URL that has been (or could have been) returned as the top-level `"url":"..."` field of the payload of `GET`.
+
+  Generally, the user is expected to retrieve the value of the object under the key `"data"` of the top-level payload of a `GET` request, alter this object, and send its modified version back to the server.
+
+  `PATCH` respects the format of the `"data"` controlled by the `&format=` URL parameter. Sending a `PATCH` request to the URL with certain `&format=X` can only modify the fields returned by the `GET` request of this `&format=X`.
+
+  For `PATCH` requests, the key of the record being updated is taken from the URL. The user is expected to either not send in the key as part of the payload, or ensure the key contained in the payload matches the key contained in the URL.
+  
+  `PATCH` will fail if attempted to modify a non-existing resource, resulting in `404 (Not Found)` error code.
+  
+  Also, `PATCH` performs conflict resolution. The URL being `PATCH`-ed must contain a revision / timestamp of the resource. The exact semantics of the revision / timestamp is not specified by this RFC, but the top-level `"url"` field returned by the `GET` request is guaranteed to have it set.
+
+  An attempt to `PATCH` to a URL that points to the version of the resource other than the most recent one will result in no data mutation and a `409 Conflict` error code. The user is then expected to re-run the `GET` request, followed by the re-run `PATCH` request onto the URL of the most recent version of the resource returned by the `GET` request. Confirming that the modifications that occurred in the meantime do not conflict with the new payload being sent in the body of the `PATCH` request is then the responsibility of the API user.
+
+* **PUT**: Create a new record for a resource, or update all fields of an existing resource.
+
+  Attempting to `PUT` to a URL that has an incomplete view of a record would result in a `405 Method Not Allowed` error.
+  
+  In the `&format=full` view, `PUT` is equivalent to `PATCH`. 
+  
+  `PUT` behaves the same way as `PATCH` when it comes to avoiding merge conflicts.
+
+  When legal, the only difference between `PUT` and `PATCH` is that `PUT` will create a new resource if it did not exist before the call. A `201 Created` response code will then be returned. Use `PATCH` if this behavior is undesirable.
```

Comment:
Of course -- terminology issue from our end, my bad.

Say, for `/user` resources, our (internal) way of thinking of `PUT` and `POST` implies assuming `/user` and `/user/123` are actually the same endpoint -- we effectively treat `/123` as a URL argument for `/user`, as `/user` is registered as "a route accepting zero or one slash-separated URL arguments after itself".

Agree it should be phrased cleaner. Thanks!